app/controllers/tools_controller.rb

```diff
@@ -123,7 +123,34 @@ def email
       @actionPage = @action_page
       render "email_target"
     else
-      redirect_to @action_page.email_campaign.service_uri(params[:service])
+      if params[:state_rep_email]
+        redirect_to @action_page.email_campaign.service_uri(params[:service], { email: params[:state_rep_email] })
+      else
+        redirect_to @action_page.email_campaign.service_uri(params[:service])
+      end
+    end
+  end
+
+  # GET /tools/state_reps
+  #
+  # This endpoint is hit by the js for state legislator lookup-by-address actions.
+  # It renders json containing html markup for presentation on the view
+  def state_reps
+    @email_campaign = EmailCampaign.find(params[:email_campaign_id])
+    @actionPage = @email_campaign.action_page
+    address = "#{params[:street_address]} #{params[:zipcode]}"
+    civic_api_response = CivicApi.state_rep_search(address, @email_campaign.leg_level)
+    @state_reps = JSON.parse(civic_api_response.body)["officials"]
+    state_rep_emails = []
+    @state_reps.each do |sr|
+      state_rep_emails << sr["emails"] if !sr["emails"].nil?
+    end
+    # single-rep lookup only
+    @state_rep_email = state_rep_emails.flatten.first
```

Comment:
this might be overkill but it might be nice to have a setup like this:

- `StateLevelInfo` - Service object type class responsible for turning info from the controller into an array of state representative objects
- `CivicApi` - Module responsible for making the API calls themselves (pretty much what we have now)
- `StateRep` - Tiny class that makes nice state rep objects from JSON

```ruby
class StateLevelInfo
  VALID_ROLES = %w(legislatorLowerBody legislatorUpperBody
                   headOfGovernment).freeze

  def self.reps_for(**args)
    new(**args).reps
  end

  def new(leg_level:, address:)
    @role = leg_level
    @address = address
  end

  def reps
    return unless valid_role?
    return unless address.present?
    return unless api_response.present?
    # might need another guard clause here to check that we have a real,
    # non-error response?
    StateRep.parse_from_json(api_response)
  end

  private

  attr_reader :role, :address

  def valid_role?
    # could potentally do this in the CivicApi module instead,
    # in which case this class only needs to check that the role is present
    VALID_ROLES.include? role
  end

  def api_response
    @response ||= CivicApi.state_rep_search(address, role).try(:body)
  end
end

class StateRep
  attr_reader :name, :emails

  def self.parse_from_json(json_response)
    JSON.parse(json_response)["officials"].map do |h|
      rep = h.symbolize_keys.slice(%i(name emails)) 
      rep[:emails] = rep.emails.flatten # might not be necessary
      new(**rep)
    end
  end

  def new(name:, emails:)
    @name = name
    @emails = emails
  end

  def info_text
    str = "Your representative is #{name}.\n"
    str << if emails.present?
             "They can be reached at #{emails.join(", ")}"
           else
             "We could not find their email address."
           end
    str
  end
end

# In controller

def state_reps
  @email_campaign = EmailCampaign.find(params[:email_campaign_id])
  @actionPage = @email_campaign.action_page
  address = address_params.to_h.values.join(" ")
  @state_reps = StateLevelInfo.reps_for(role: @email_campaign.leg_level,
                                        address: address)
  @state_rep_email = @state_reps.first.try(:emails)
  if @state_reps.present?
    render json: { content: render_to_string(partial: "action_page/state_reps") }, status: 200
  else
    render json: { error: "No representatives found" }, status: 200
  end
end
```
```erb
<div class="state-rep-names-and-emails">
  This action is for the <%= legislative_level_from_state_representative_info(@email_campaign.leg_level)} %>.
  <% @state_reps.each do |sr| -%>
    <div>
      <%= sr.info_text %>
    </div>
  <% end %>
</div>
```

We could even include the `CivicApi` module in the first class if we wanted.

Reply:
Yes! Feel free to turn this into its own issue